Cannot find module '../../../material-ui-styles/src' or its corresponding type declarations.

<!-- Provide a general summary of the issue in the Title above -->

<!--
  Thank you very much for contributing to Material-UI by creating an issue! ❤️
  To avoid duplicate issues we ask you to check off the following list.
-->

<!-- Checked checkbox should look like this: [x] -->

- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

The typings of components inside the lab are referencing `import("../../../material-ui-styles/src")`. The components inside the core are fine, so likely not something with our setup.

<!-- Describe what happens instead of the expected behavior. -->

## Expected Behavior 🤔

The typings of components in the lab should be referencing `@material-ui/styles`.

<!-- Describe what should happen. -->

## Steps to Reproduce 🕹

<!--
  Provide a link to a live example (you can use codesandbox.io) and an unambiguous set of steps to reproduce this bug.
  Include code to reproduce, if relevant (which it most likely is).

  You should use the official codesandbox template as a starting point: https://material-ui.com/r/issue-template

  If you have an issue concerning TypeScript please start from this TypeScript playground: https://material-ui.com/r/ts-issue-template

  Issues without some form of live example have a longer response time.
-->

```json
{
  "compilerOptions": {
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "baseUrl": "src",
    "declaration": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "jsx": "react",
    "lib": ["dom", "dom.iterable", "esnext"],
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmitOnError": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "pretty": true,
    "sourceMap": true,
    "strict": true,
    "target": "es2020",
    "resolveJsonModule": true,
    "rootDir": "src",
    "outDir": "build",
    "incremental": true
  },
  "include": ["src/**/*"],
  "exclude": ["**/*.spec.ts", "**/*.spec.tsx"]
}
```

Steps:

1. Add a `DatePicker` form the lab
2. Build your project using  `tsc`
3. Get error `Cannot find module '../../../material-ui/src' or its corresponding type declarations.`

## Context 🔦

<!--
  What are you trying to accomplish? How has this issue affected you?
  Providing context helps us come up with a solution that is most useful in the real world.
-->

## Your Environment 🌎

<!--
  Run `npx @material-ui/envinfo` and post the results.
  If you encounter issues with TypeScript please include the used tsconfig.
-->
<details>
  <summary>`npx @material-ui/envinfo`</summary>
  
```
System:
    OS: macOS 11.1
  Binaries:
    Node: 14.15.1 - ~/.nvm/versions/node/v14.15.1/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.8 - ~/.nvm/versions/node/v14.15.1/bin/npm
  Browsers:
    Chrome: 87.0.4280.88
    Edge: 87.0.664.66
    Firefox: Not Found
    Safari: 14.0.2
  npmPackages:
    @emotion/react: ^11.1.4 => 11.1.4 
    @emotion/styled: ^11.0.0 => 11.0.0 
    @material-ui/core: ^5.0.0-alpha.20 => 5.0.0-alpha.20 
    @material-ui/icons: ^5.0.0-alpha.20 => 5.0.0-alpha.20 
    @material-ui/lab: ^5.0.0-alpha.20 => 5.0.0-alpha.20 
    @material-ui/styled-engine:  5.0.0-alpha.20 
    @material-ui/styles:  5.0.0-alpha.20 
    @material-ui/system:  5.0.0-alpha.20 
    @material-ui/types:  5.1.2 
    @material-ui/unstyled:  5.0.0-alpha.20 
    @material-ui/utils:  5.0.0-alpha.20 
    @types/react: ~17 => 17.0.0 
    react: ~17 => 17.0.1 
    react-dom: ~17 => 17.0.1 
    typescript: ^4.1.3 => 4.1.3 
```
</details>

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -209,14 +209,8 @@ jobs:
             git --no-pager diff
 
       - run:
-          name: Log defect declaration files
-          command: |
-            # ignore build failures
-            # Fixing these takes some effort that isn't viable to merge in a single PR.
-            # We'll simply monitor them for now.
-            set +e
-            node scripts/testBuiltTypes.js
-            exit 0
+          name: Any defect declaration files?
+          command: node scripts/testBuiltTypes.js
       - save_cache:
           name: Save generated declaration files
           key: typescript-declaration-files-{{ .Branch }}-{{ .Revision }}

--- a/packages/material-ui-lab/package.json
+++ b/packages/material-ui-lab/package.json
@@ -29,7 +29,7 @@
     "build:node": "node ../../scripts/build node",
     "build:stable": "node ../../scripts/build stable",
     "build:copy-files": "node ../../scripts/copy-files.js",
-    "build:types": "tsc -p tsconfig.build.json",
+    "build:types": "node ../../scripts/buildTypes",
     "prebuild": "rimraf build",
     "release": "yarn build && npm publish build --tag next",
     "test": "cd ../../ && cross-env NODE_ENV=test mocha 'packages/material-ui-lab/**/*.test.{js,ts,tsx}'",

--- a/packages/material-ui-utils/package.json
+++ b/packages/material-ui-utils/package.json
@@ -28,7 +28,7 @@
     "build:node": "node ../../scripts/build node",
     "build:stable": "node ../../scripts/build stable",
     "build:copy-files": "node ../../scripts/copy-files.js",
-    "build:types": "tsc -p tsconfig.build.json",
+    "build:types": "node ../../scripts/buildTypes",
     "prebuild": "rimraf build",
     "release": "yarn build && npm publish build --tag next",
     "test": "cd ../../ && cross-env NODE_ENV=test mocha 'packages/material-ui-utils/**/*.test.{js,ts,tsx}'",

--- a/packages/material-ui/package.json
+++ b/packages/material-ui/package.json
@@ -33,7 +33,7 @@
     "build:stable": "node ../../scripts/build stable",
     "build:umd": "cross-env BABEL_ENV=stable rollup -c scripts/rollup.config.js",
     "build:copy-files": "node ../../scripts/copy-files.js",
-    "build:types": "tsc -p tsconfig.build.json",
+    "build:types": "node ../../scripts/buildTypes",
     "extract-error-codes": "cross-env MUI_EXTRACT_ERROR_CODES=true yarn build:modern",
     "prebuild": "rimraf build tsconfig.build.tsbuildinfo",
     "release": "yarn build && npm publish build --tag next",

--- a/scripts/buildTypes.js
+++ b/scripts/buildTypes.js
@@ -1,0 +1,133 @@
+const chalk = require('chalk');
+const childProcess = require('child_process');
+const glob = require('fast-glob');
+const fse = require('fs-extra');
+const path = require('path');
+const { promisify } = require('util');
+const yargs = require('yargs');
+
+const exec = promisify(childProcess.exec);
+
+/**
+ * Fixes a wrong import path caused by https://github.com/microsoft/TypeScript/issues/39117
+ *
+ * @remarks Paths are hardcoded since it is unclear if all these broken import paths target "public paths".
+ *
+ * @param {string} importPath
+ */
+function rewriteImportPath(importPath) {
+  if (importPath.startsWith('../material-ui/src')) {
+    return importPath.replace('../material-ui/src', '@material-ui/core');
+  }
+  if (importPath.startsWith('../material-ui-styles/src')) {
+    return importPath.replace('../material-ui-styles/src', '@material-ui/styles');
+  }
+
+  throw new Error(`Don't know where to rewrite '${importPath}' to`);
+}
+
+async function main() {
+  const packageRoot = process.cwd();
+
+  const tsconfigPath = path.join(packageRoot, 'tsconfig.build.json');
+  if (!fse.existsSync(tsconfigPath)) {
+    throw new Error(
+      'Unable to find a tsconfig to build this project. ' +
+        `The package root needs to contain a 'tsconfig.build.json'. ` +
+        `The package root is '${packageRoot}'`,
+    );
+  }
+
+  await exec(['yarn', 'tsc', '-p', tsconfigPath].join(' '));
+
+  const publishDir = path.join(packageRoot, 'build');
+  const declarationFiles = await glob('**/*.d.ts', { absolute: true, cwd: publishDir });
+  if (declarationFiles.length === 0) {
+    throw new Error(`Unable to find declaration files in '${publishDir}'`);
+  }
+
+  async function rewriteImportPaths(declarationFile) {
+    const code = await fse.readFile(declarationFile, { encoding: 'utf8' });
+    let fixedCode = code;
+    const changes = [];
+
+    // find all type `import()`
+    // not to be confused with `import type`
+    const importTypeRegExp = /import\(([^)]+)\)/g;
+
+    let importTypeMatch;
+    // eslint-disable-next-line no-cond-assign -- Waiting for RegExp.prototype.matchAll
+    while ((importTypeMatch = importTypeRegExp.exec(code)) !== null) {
+      // First and last character are quotes.
+      // TypeScript mixes single and double quotes.
+      const importPath = importTypeMatch[1].slice(1, -1);
+      // In filesystem semantics `@material-ui/core` is a relative path.
+      // But when resolving imports these specifiers are considered "bare specifiers" and work differently.
+      // We're only interested in imports that are considered "relative path imports".
+      const isBareImportSpecifier = !importPath.startsWith('.');
+      if (!isBareImportSpecifier) {
+        const resolvedImport = path.resolve(declarationFile, importPath);
+        const importPathFromPublishDir = path.relative(publishDir, resolvedImport);
+        const isImportReachableWhenPublished = !importPathFromPublishDir.startsWith('.');
+
+        if (!isImportReachableWhenPublished) {
+          try {
+            const fixedImportPath = rewriteImportPath(importPathFromPublishDir);
+            const originalImportType = importTypeMatch[0];
+            const fixedImportType = importTypeMatch[0].replace(importPath, fixedImportPath);
+
+            // Make it easy to visually scan for the created lines.
+            changes.push(`-${chalk.bgRed(originalImportType)}\n+${chalk.bgGreen(fixedImportType)}`);
+            fixedCode = fixedCode.replace(originalImportType, fixedImportType);
+          } catch (error) {
+            throw new Error(`${declarationFile}: ${error}`);
+          }
+        }
+      }
+    }
+
+    const changed = changes.length > 0;
+    if (changed) {
+      await fse.writeFile(declarationFile, fixedCode);
+    }
+
+    return changes;
+  }
+
+  let rewrittenTally = 0;
+  let errorTally = 0;
+  await Promise.all(
+    declarationFiles.map(async (declarationFile) => {
+      try {
+        const rewrites = await rewriteImportPaths(declarationFile, publishDir);
+        if (rewrites.length > 0) {
+          // eslint-disable-next-line no-console -- Verbose logging
+          console.log(`${chalk.bgYellow`FIXED`} '${declarationFile}':\n${rewrites.join('\n')}`);
+          rewrittenTally += 1;
+        } else {
+          // eslint-disable-next-line no-console -- Verbose logging
+          console.log(`${chalk.bgGreen`OK`} '${declarationFile}'`);
+        }
+      } catch (error) {
+        console.error(error);
+        errorTally += 1;
+        process.exitCode = 1;
+      }
+    }),
+  );
+
+  // eslint-disable-next-line no-console -- Verbose logging
+  console.log(`Fixed: ${rewrittenTally}\nFailed: ${errorTally}\nTotal: ${declarationFiles.length}`);
+}
+
+yargs
+  .command({
+    command: '$0',
+    description:
+      'Builds a project with a fix for https://github.com/microsoft/TypeScript/issues/39117',
+    handler: main,
+  })
+  .help()
+  .strict(true)
+  .version(false)
+  .parse();

--- a/scripts/buildTypes.js
+++ b/scripts/buildTypes.js
@@ -16,11 +16,14 @@ const exec = promisify(childProcess.exec);
  * @param {string} importPath
  */
 function rewriteImportPath(importPath) {
-  if (importPath.startsWith('../material-ui/src')) {
-    return importPath.replace('../material-ui/src', '@material-ui/core');
+  const coreSrcPath = path.join('..', 'material-ui', 'src');
+  if (importPath.startsWith(coreSrcPath)) {
+    return importPath.replace(coreSrcPath, '@material-ui/core');
   }
-  if (importPath.startsWith('../material-ui-styles/src')) {
-    return importPath.replace('../material-ui-styles/src', '@material-ui/styles');
+
+  const stylesSrcPath = path.join('..', 'material-ui-styles', 'src');
+  if (importPath.startsWith(stylesSrcPath)) {
+    return importPath.replace(stylesSrcPath, '@material-ui/styles');
   }
 
   throw new Error(`Don't know where to rewrite '${importPath}' to`);

--- a/scripts/buildTypes.js
+++ b/scripts/buildTypes.js
@@ -13,15 +13,15 @@ const exec = promisify(childProcess.exec);
  *
  * @remarks Paths are hardcoded since it is unclear if all these broken import paths target "public paths".
  *
- * @param {string} importPath
+ * @param {string} importPath - POSIX path
  */
 function rewriteImportPath(importPath) {
-  const coreSrcPath = path.join('..', 'material-ui', 'src');
+  const coreSrcPath = path.posix.join('..', 'material-ui', 'src');
   if (importPath.startsWith(coreSrcPath)) {
     return importPath.replace(coreSrcPath, '@material-ui/core');
   }
 
-  const stylesSrcPath = path.join('..', 'material-ui-styles', 'src');
+  const stylesSrcPath = path.posix.join('..', 'material-ui-styles', 'src');
   if (importPath.startsWith(stylesSrcPath)) {
     return importPath.replace(stylesSrcPath, '@material-ui/styles');
   }
@@ -75,7 +75,10 @@ async function main() {
 
         if (!isImportReachableWhenPublished) {
           try {
-            const fixedImportPath = rewriteImportPath(importPathFromPublishDir);
+            const fixedImportPath = rewriteImportPath(
+              // ensure relative POSIX path
+              importPathFromPublishDir.replace(/\\/g, '/'),
+            );
             const originalImportType = importTypeMatch[0];
             const fixedImportType = importTypeMatch[0].replace(importPath, fixedImportPath);
 

